doc/remediation-templating.md

```diff
@@ -0,0 +1,207 @@
+XCCDF values from rules. 
+
+The following example is parts of two YAML files from ComplianceAsCode content, 
+the first one is the rule, and the second is the remediation:
+```yaml
+documentation_complete: true
+title: 'Set SSH Idle Timeout Interval'
+...
+requires:
+    - sshd_set_keepalive
+
+ocil_clause: 'it is commented out or not configured properly'
+
+ocil: |-
+    Run the following command to see what the timeout interval is:
+    <pre>$ sudo grep ClientAliveInterval /etc/ssh/sshd_config</pre>
+    If properly configured, the output should be:
+    <pre>ClientAliveInterval {{{ xccdf_value("sshd_idle_timeout_value") }}}</pre>
+...
+```
+As showing in the rule YAML file above, an XCCDF value for SSH Idle Timeout Interval is set.
+
+In the remediation YAML file, we can modify that Marco to pass the name of the 
+XCCDF value `VAR_SSHD_IDLE_TIMEOUT_VALUE`, and then modify the related `ClientAliveCountMax`
+values to `VAR_SSHD_IDLE_TIMEOUT_VALUE`:
+```yaml
+---
+# platform = multi_platform_ocp,multi_platform_rhcos
+# reboot = false
+# strategy = restrict
+# complexity = low
+# disruption = low
+{{{ kubernetes_sshd_set("VAR_SSHD_IDLE_TIMEOUT_VALUE") }}}
+```
+
+And the machine config object data will be something like below before it gets url-encoded.
+```jinja
+...
+PrintLastLog yes
+#TCPKeepAlive yes
+PermitUserEnvironment no
+Compression no
+ClientAliveInterval 600
+ClientAliveCountMax VAR_SSHD_IDLE_TIMEOUT_VALUE
+
+#UseDNS no
+#PidFile /var/run/sshd.pid
+#MaxStartups 10:30:100
+#PermitTunnel no
+...
+```
+Since the url encode doesn't encode ASCII alpha-numerical nor underscore characters, and
+we can later change the place holder after the url encoding from `VAR_SSHD_IDLE_TIMEOUT_VALUE`
+to `{{sshd_idle_timeout_value}}`, so that it can be post-process them with the appropriate
+escaped instances The following code is the example of final result:
+```xml
+<fix rule="sshd_set_idle_timeout" complexity="low" disruption="low" reboot="false" strategy="restrict">---
+apiVersion: machineconfiguration.openshift.io/v1
+kind: MachineConfig
+spec:
+  config:
+    ignition:
+      version: 3.1.0
+    storage:
+      files:
+      - contents:
+          source: data:, `PrintLastLog%20yes%0A%23TCPKeepAlive%20yes%0APermitUserEnvironment%20no%0ACompression
+          %20no%0AClientAliveInterval%20600%0AClientAliveCountMax%20{{sshd_idle_timeout_value}}%0A%0A%23UseDNS%2
+          0no%0A%23PidFile%20%2Fvar%2Frun%2Fsshd.pid%0A%23MaxStartups%2010%3A30%3A100%0A%23PermitTunnel%20no`
+        mode: 0600
+        path: /etc/ssh/sshd_config
+        overwrite: true
+</fix>
+```
+
+There will be some change in the aggregator since we brought a marker XCCDF variable into
+the remediation content, we need to replace all the placeholder with corresponding values
+before we create the `ComplianceRemediation` object. we need to fetch all the variables that 
+are in the scan result `ConfigMap` for all the nodes with consistency check(Note that all the 
```

Comment:
The variable values also need to be url-encoded right?

Reply:
Yes, the corresponding values needed to be url-encoded as well before we substitute to the XCCDF value marker.

Reply:
This could be done via golang templating https://golang.org/pkg/text/template/#hdr-Functions so we could safely add some functions to the template and the Compliance Operator could evaluate them